vsphere/resource_vsphere_license.go

```diff
@@ -0,0 +1,246 @@
+package vsphere
+
+import (
+	"errors"
+	"fmt"
+	"log"
+
+	"context"
+
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/vmware/govmomi"
+	"github.com/vmware/govmomi/license"
+	"github.com/vmware/govmomi/vim25/methods"
+	"github.com/vmware/govmomi/vim25/types"
+)
+
+var (
+	// ErrNoSuchKeyFound is an error primarily thrown by the Read method of the resource.
+	// The error doesn't display the key itself for security reasons.
+	ErrNoSuchKeyFound = errors.New("The key was not found")
+	// ErrKeyCannotBeDeleted is an error which occurs when a key that is used by VMs is
+	// being removed
+	ErrKeyCannotBeDeleted = errors.New("The key wasn't deleted")
+)
+
+func resourceVSphereLicense() *schema.Resource {
+	return &schema.Resource{
+
+		SchemaVersion: 1,
+
+		Create: resourceVSphereLicenseCreate,
+		Read:   resourceVSphereLicenseRead,
+		Update: resourceVSphereLicenseUpdate,
+		Delete: resourceVSphereLicenseDelete,
+
+		Schema: map[string]*schema.Schema{
+			"license_key": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"labels": &schema.Schema{
+				Type:     schema.TypeMap,
+				Optional: true,
+			},
+
+			// computed properties returned by the API
+			"edition_key": &schema.Schema{
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"name": &schema.Schema{
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"total": &schema.Schema{
+				Type:     schema.TypeInt,
+				Computed: true,
+			},
+			"used": &schema.Schema{
+				Type:     schema.TypeInt,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceVSphereLicenseCreate(d *schema.ResourceData, meta interface{}) error {
+	log.Println("[INFO] Running the create method")
+
+	client := meta.(*govmomi.Client)
+	manager := license.NewManager(client.Client)
+
+	key := d.Get("license_key").(string)
+
+	log.Println(" [INFO] Reading the key from the resource data")
+	finalLabels := make(map[string]string)
+
+	if labels, ok := d.GetOk("labels"); ok {
+		labelMap := labels.(map[string]interface{})
+		for key, value := range labelMap {
+			finalLabels[key] = value.(string)
+		}
+	}
+
+	var info types.LicenseManagerLicenseInfo
+	var err error
+	switch t := client.ServiceContent.About.ApiType; t {
+	case "HostAgent":
+		// Labels are not allowed in ESXi
+		if len(finalLabels) != 0 {
+			return errors.New("Labels are not allowed in ESXi")
+		}
+
+		info, err = manager.Update(context.TODO(), key, finalLabels)
```

Comment:
Since labels can't be used on ESXi period, it might be good to just remove `finalLabels` here and just use `nil`.